smb/src/resource.rs

```diff
@@ -238,7 +239,13 @@ pub struct ResourceHandle {
     name: String,
     handler: HandlerReference<ResourceMessageHandle>,
 
-    file_id: FileId,
+    // Whether the resource is open or not.
+    // TODO: Consider using RwLock here on FileId instead of AtomicBool+FileId.
+    open: AtomicBool,
+
+    // Avoid accessing directly; use the `file_id()` getter,
+    // that makes sure the resource is still open.
+    _file_id: FileId,
```

Comment:
So, from my reading of this, the only reason open needs to be atomic is because `close` takes `&self` instead of `&mut self`. Rust won't let you send a & or &mut to another thread. This means you only have a single owner of a `Resource`. Any sharing of that `Resource` must be through an std::sync::Arc (or something similar). Any mutable sharing will have to be through `Arc<SyncPrimitive>` where `SyncPrimitive` provides `Sync` such as an async Mutex/RwLock/etc.

Can we make `close` just take a `&mut self` and then make `open` a bool? `Drop` already takes &mut self so you're good there. Of course, let me know if I'm missing something.